Added an option to commit without hooks

koraktor opened the pull request:
While writing a [blog post](http://koraktor.posterous.com/continuous-integration-using-git) about Git as a continuous integration tool, I noticed that GitX doesn't have an option to skip the hooks called before a commit is done. This hooks are `pre-commit` and `commit-msg` and GitX calls them always. This way you will never be able to commit with GitX when one of your hooks fails (for whatever reason that may be). With command-line Git you have `git commit --no-verify` which disable these hooks. I added a small checkbox below the commit message field (next to the amend checkbox) for this purpose.

epall committed:
Allow zero lines of context

Hunk/line patches don't generally have context

Before this, every attempt to stage a hunk or line was met with
"patch failed" using git 1.6.6

Commit:
Fix a freeze when staging/unstaging thousands of files

nriley committed:
Don't remove a file from "Unstaged Changes" when attempting to discard changes if it is a new file; it'll just reappear when you refresh.

nriley committed:
Fix vertical alignment of multiline Subjects in history.

treelinelabs committed:
Display output of pre-commit hook

mdippery committed:
.gitignore: Ignore everything but Xcode project settings

brotherbard committed:
Fix "Duplicate declaration of method" error when using LLVM

The class method and the instance method had different types for the window controller. The instance method should be using the NSWindowController since the sheet can be shown in a non PBGitWindowController window (the PBCloneRepositoryPanel).

brotherbard committed:
Fix issues reported by Xcode's build and analyze

Using the LLVM GCC 4.2 compiler.

brotherbard committed:
Fix crashes from NSInvalidArgumentException

This happens when running the gitx cli tool from a directory that does not contain a git repository.

koraktor committed:
Don't follow symlinks when setting Git path

This allows using /usr/local/bin/git (using Homebrew) to work across updates

Commit:
Fixing compiler warnings.

liyanage committed:
Add "Copy SHA" alternate menu item to main menu, with default binding Cmd-Option-C. It copies just the leading 7 characters of the SHA instead of the whole commit info.

brotherbard committed:
Fix spelling error

brotherbard committed:
Fix getting the correct git directory

If the path already includes the .git directory then rev-parse will return "." to mean the current directory.

brotherbard committed:
Include the CFBundleGitVersion in the version command of the CLI

CFBundleGitVersion has the git description information in it.

brotherbard committed:
Refactor the gitx CLI to use apple events and the scripting bridge

Sending the arguments with the openURL:... message allows the repository document to modify it's UI without the UI flashing between states as it opens.

Covers all the existing functionality of the CLI, but modifies:
    - "--all" "--local" "--branch" change the branch filter
    - cleaned up the usage (help) text and added info on missing commands
    - looks up the full ref name of refs so the name of a branch or tag can be entered (the user can enter "master" instead of "refs/heads/master")

Modified the History Controller to watch for and react to branch filter changes.

The GitX.h file is generated by the 'sdp' tool in a run script build phase called 'Generate Scripting Bridge Header' based on the content of GitX.sdef. It is used by the Scripting Bridge so that other apps (in this case the gitx CLI) can call Applescript commands on GitX in objective-c.

brotherbard committed:
Add "--git-dir=" to gitx CLI

Allows specifying a repository in a particular location for the gitx operation.

brotherbard committed:
Add --init and --clone to gitx CLI and Applescript

Allows creating repositories from the command line, Applescript, or the scripting bridge. These are basic commands, if you need to use commandline options then use git itself.

brotherbard committed:
Don't allow drag-n-drop for the checked out branch.

Moving the checked out branch would cause odd things to happen to git's index.

brotherbard committed:
Add a relative date column to the commits list table view

Shows a relative time ("x minutes ago", "Yesterday", "x days ago", "x years ago").
    - Sort the column on the commit's date and not on the string value
    - Use a custom formatter connected to the cell in the XIB
    - Show or hide using the contextual menu for the table header

brotherbard committed:
Fix sorting of the subject column.

Was broken back in 4e4ffcc.

brotherbard committed:
Improved sidebar icons for Branch, Remote Branch and Tags

Icons by Thorsten Seitz <thorsten@portableinnovations.de>

brotherbard committed:
Improve search UI

Previously searching would filter the commits in the commit tableview to only show the commits that matched the search. However the context of where those commits exist in the history is lost.

With this patch all the commits are shown but the commits that match the search are highlighted with a light blue background. In addition there is a forward/back button to step through the matches.

A new search controller:
    - keeps track of the matching results
    - finds the next or previous result
    - displays the number of matches found or "Not found"
    - shows/hides the # of matches text and the next/last stepper button
    - shows a small bezel style window with a rewind icon indicating that the selection has cycled (pressing next when at the last match or previous when at the first)
    - sets up the search predicate which covers Subject, Author and SHA (previously this was three different searches)
    - stores search results in an NSIndexSet to make finding if a row is in the set faster (needed at drawing time)

Highlighting of search result rows is done in PBCommitList -drawRow:clipRect:

PBGitTextFieldCell is a subclass of NSTextFieldCell that disables the cell's selection highlighting.

Supporting Find Next and Find Previous (cmd-g and cmd-shift-g) menu commands required changing the action method of the menu items because NSTextFields (seem to) actively disable items in the Find menu.

rewindImage.pdf created by Nathan Kinsinger

brotherbard committed:
Remove the isReachable test for gravatar.com

If there is no network connection then the reachability test will take the system default time (90 seconds I think?) to time out. During this time the UI is blocked (spinning beach ball).

The test was being done even if gravatar use was turned off. It doesn't really matter if the site is unreachable, the image will not load but that's not a big deal.

brotherbard committed:
Add a Short SHA column to the commit list

Shows only the first 10 characters of the full SHA hash so as to take up less room. Show or hide it using the contextual menu for the table header.

brotherbard committed:
Disable deleting and pushing the detached HEAD

brotherbard committed:
Prevent drag and drop of tags and remote branches

Moving an annotated tag would convert it to a lightweight tag and lose the tags message.

Given that tags are meant to represent a fixed point in the history and remote branches represent the state of the remote I'm disabling moving them.

brotherbard committed:
Add a recent searches menu to the search field

brotherbard committed:
Add the contextual menu to the other columns of the commit list

Now all the columns have the menu for the commit, not just the subject column.

brotherbard committed:
Add search modes for git pickaxe

Add two new search modes to search the commits for strings that were added or removed in a commit.
    - match by string or an extended POSIX regex (see man git-log -S<string> and --pickaxe-regex)
    - include the modes in the search menu
    - store the current mode in user defaults
    - show progress indicator when search takes time

brotherbard committed:
Add search commands to the gitx cli

-s or --search= for searching in subject, author or SHA
-S or --Search= for git's pickaxe string matching
-r or --regex= for git's pickaxe regex matching

Commit:
Make the context slider work again

brotherbard committed:
Add a File Path search mode

Highlights any commit that touches a file.

The "Show history of file" context menu item will now set the path search instead of creating a rev specifier.

brotherbard commented:
I don't think putting a checkbox in the commit view is a good idea. Just something to confuse people.

I think a better solution would be to show an error sheet with the commit error and have a button to force the commit there. That way only people that have pre-commit hooks will have to deal with it.

See http://groups.google.com/group/gitx/browse_thread/thread/a03bcab60844b812 for a patch to show the error. Then you will need to create a sheet that has buttons to commit or not.

koraktor committed:
Also display output of commit-msg hook

koraktor commented:
Sounds reasonable. I'll take a look at that.

brotherbard committed:
Allow the stage view's context slider to move from 0 to 10

brotherbard committed:
Bugfix: include missing scripting method definitions

koraktor committed:
Added a dialog for failing commit hooks

This enables the user to force a commit (i.e. --no-verify) when either pre-commit or
commit-msg hook fails.

koraktor commented:
This one works as suggested. It might be a more general solution after some refactoring, e.g. for non-forward pushes, branch replacements etc.
